File: basic_dsa_three_hundred_forty_one.java
/**
 * Basic DSA Problems - Flatten Nested List Iterator (LeetCode 341)
 * Iterator that flattens nested list structure
 */
public class basic_dsa_three_hundred_forty_one {

    // Problem: Nested list. Implement NestedIterator with next() and hasNext().

    interface NestedInteger {
        boolean isInteger();
        Integer getInteger();
        java.util.List<NestedInteger> getList();
    }

    static class NestedInt implements NestedInteger {
        Integer val;
        java.util.List<NestedInteger> list;
        NestedInt(int v) { val = v; list = null; }
        NestedInt(java.util.List<NestedInteger> l) { val = null; list = l; }
        public boolean isInteger() { return val != null; }
        public Integer getInteger() { return val; }
        public java.util.List<NestedInteger> getList() { return list; }
    }

    static class NestedIterator {
        private java.util.Iterator<Integer> it;

        public NestedIterator(java.util.List<NestedInteger> nestedList) {
            java.util.List<Integer> flat = new java.util.ArrayList<>();
            flatten(nestedList, flat);
            it = flat.iterator();
        }

        private void flatten(java.util.List<NestedInteger> list, java.util.List<Integer> out) {
            for (NestedInteger ni : list) {
                if (ni.isInteger()) out.add(ni.getInteger());
                else flatten(ni.getList(), out);
            }
        }

        public boolean hasNext() { return it.hasNext(); }
        public Integer next() { return it.next(); }
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Flatten Nested List Iterator");
        System.out.println("=".repeat(60));

        java.util.List<NestedInteger> list = java.util.Arrays.asList(
            new NestedInt(1),
            new NestedInt(java.util.Arrays.asList(new NestedInt(2), new NestedInt(java.util.Arrays.asList(new NestedInt(3)))))
        );
        NestedIterator it = new NestedIterator(list);
        java.util.List<Integer> out = new java.util.ArrayList<>();
        while (it.hasNext()) out.add(it.next());
        System.out.println("[1,[2,[3]]] -> " + out);
    }
}

File: basic_dsa_three_hundred_forty_one.cpp
/**
 * Basic DSA - Implement Stack using Queues (LeetCode 225)
 * push O(n), top/pop O(1) using two queues.
 */
#include <iostream>
#include <string>
#include <queue>

class MyStack {
    std::queue<int> q1, q2;
public:
    void push(int x) {
        q2.push(x);
        while (!q1.empty()) {
            q2.push(q1.front());
            q1.pop();
        }
        std::swap(q1, q2);
    }
    int pop() {
        int v = q1.front();
        q1.pop();
        return v;
    }
    int top() { return q1.front(); }
    bool empty() { return q1.empty(); }
};

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Implement Stack using Queues\n";
    std::cout << std::string(60, '=') << "\n";

    MyStack st;
    st.push(1);
    st.push(2);
    std::cout << "top -> " << st.top() << "\n";
    std::cout << "pop -> " << st.pop() << "\n";
    std::cout << "empty -> " << (st.empty() ? "true" : "false") << "\n";
}
